### ISSUE TITLE
Character count

### ISSUE THREAD

**Issue by govuk-design-system** [alphagov/govuk-design-system-backlog#67]
Use this issue to discuss the [character count component](https://design-system.service.gov.uk/components/character-count/).

**Comment by tombye** [comment]
@fofr does your implementation include messaging for when users go over the number of words?

Curious because I've had that in previous projects but not sure it was ever validated as a need.

**Comment by fofr** [comment]
@tombye Yes, we've used the messaging that was in the original repo by @alex-ju. The messaging has worked well when users have pasted in text that was too long from another document.

![screen shot 2018-08-07 at 14 15 50](https://user-images.githubusercontent.com/319055/43778021-6b6256c4-9a4c-11e8-951f-7f8c107d3530.png)

**Comment by timpaul** [comment]
We're thinking about the possible error states for this component. In particular, how the 'You have 26 words too many' error should be represented in the [error summary component](https://design-system.service.gov.uk/components/error-summary/) that can appear at the top of a page.

Just repeating that message wouldn't give context about _which_ form field the error related to. As the example above from @fofr shows it's quite possible to have multiple character count components on a single screen - so, should the error be adjusted to include information about the specific field. For example "Interview process summary has too many words".

However, if we only adjusted the error message in the summary then this would go against our [advice to keep them the same](https://design-system.service.gov.uk/components/error-message/#be-consistent).

@stevenaproctor - do you have a take on this?

**Comment by stevenaproctor** [comment]
@timpaul You could use the standard too long error message ‘[whatever it is] must be {x} [characters or words] or less’. Or, if it is available, you could use the number of characters you are over in the error message ‘[whatever it is] has {x} [characters or words] too many’.

For the example, this would give you the errors:

* Interview process must be 250 words or less
* Interview process has 26 words too many

What do you think?

**Comment by nubz** [comment]
We have a need for a character count component in one of our services at HMRC. I noticed the github readme does not recommend using this component as-is in production - does anyone know what the blocker to recommendation is?

**Comment by fofr** [comment]
Something new that we've observed in user research:

We asked users to prepare content before our service was available. We gave them word count limits. We've found that users wrote their content in Word and used Word's word count feature to check the length. When they’ve then copied that content into our tool there's been a disparity between the two counts: ours is a little stricter and users have had to edit their content to reduce by 1 or 2 words. This has caused frustration, especially when the previous word document has been reviewed/approved.

Presumably the logic for word counts in Word is a little more complex than the regex used in https://github.com/alphagov/ds-character-count. I doubt that there is consistent logic between different word processors, I like the simplicity of the regex used. (It's complicated a little by our use of markdown, where `* list item` counts as three words while if you type that in Word the asterisk gets converted to a bullet point and isn't counted.)

We still need to investigate the cause more thoroughly, but wanted to share this gotcha first.

**Comment by alex-ju** [comment]
valuable found @fofr. thanks a lot for sharing it with us.

**Comment by hooperstu** [comment]
Has anybody tested this with screen readers yet?

**Comment by alex-ju** [comment]
@hooperstu it was tested with screen readers and had a few user research rounds: https://github.com/alphagov/govuk-design-system/wiki/Character-count-testing-and-user-research

The only issue I'm aware of is with JAWS who repeats field's label every time the count message is played.

**Comment by stevenaproctor** [comment]
@alex-ju I tested it today with NVDA and Firefox and the remaining characters information was read out twice, which I did not expect.

For example, when I entered the box and typed 'a' I heard 'You have 199 characters remaining' twice. When I deleted the 'a', I heard 'You have 200 characters remaining' twice as well.

Any ideas?

**Comment by alex-ju** [comment]
@stevenaproctor: thanks for reporting it. The [prototype](https://govuk-charcount-prototype.herokuapp.com/components/char-count) doesn't seem to have this issue. I will try to investigate what's causing the issue and get back to you.

**Comment by alex-ju** [comment]
Hi @stevenaproctor, I did a bit of investigation around the NVDA issue, which seemed to be caused by the polling events. Updated the code in the [Design System branch](https://deploy-preview-543--govuk-design-system-preview.netlify.com/components/character-count/) and tested again on NVDA. Please let me know if now is working as expected.

**Comment by stevenaproctor** [comment]
@alex-ju It is working as expected, thanks for taking a look.

If JavaScript is not available, is it possible to show "You have 200 characters remaining" rather than nothing?

**Comment by alex-ju** [comment]
@stevenaproctor having a default message for non-JS experience sounds like a good idea; I'll look into it. But in this case the message should maybe mention the limit instead of the remaining characters

**Comment by stevenaproctor** [comment]
@alex-ju Definitely. We tend to use something like "You can enter up to 200 characters".

**Comment by amyhupe** [comment]
## Working Group Review

This proposal was reviewed by a panel of designers from GDS, HMRC, DWP, EA and Home Office on Wednesday, 26th September 2018

The panel agreed that the pattern should be published in the GOV.UK Design System.

The panel also made the following recommendations:

#### Pre-publication

- [x] Update component description to clarify that it doesn’t set the limit, but exposes the limit to users and lets them type over it before submitting
- [x] Add guidance to explain why the count message appears below the textarea
Say that this component shouldn’t be used to enforce limits that stem from technical constraints
- [x] Get content peer reviewed before publication
- [x] Explain why both the count message and a validation error message appear when an error occurs.
- [x] Provide a default, non-dynamic message explaining the character or word limit for when JavaScript is not available
- [x] Fix behavioural issue that occurs when using component with NVDA and Firefox

#### Future work

- [ ] Consider making the count rules closer to the Microsoft Word count (see this issue for a fuller explanation)
- [ ] Add more information about research with users with access needs as it becomes available
- [ ] Look into fixing JAWS screen reader bug
- [ ] Explore how this works for users with low reading ages

**Comment by alex-ju** [comment]
Thank you all for all the help to get this done, from research, code to guidance. As I mentioned in the working group, most - if not all - the credits for this work should go to @edwardhorsford.

**Comment by edwardhorsford** [comment]
Great job team!

Credit also to @alex-ju, Per, @hannalaakso, @alicenoakes and everyone else on the patterns team.

**Comment by davidjamesstone** [comment]
Congrats to all involved in this. It's a nice, useful feature.

I do have some concerns about the following advice though:

"Do not impose character limits unless there is a user need" and "do your best to remove the limit in backend systems"

IMO _every_ text field/text area field should have a character limit. (Whether you choose to display that limit using this character count component is a different matter of course)

**Why?**
I don't want users entering war and peace into field that is meant for a full name.
I don't want an essay when a short description is expected.

The 'maxlength' attribute is your first line of defence against this kind of behaviour (which of course should be backed up with server-side validation ideally at both the route endpoint _and_ database schema level).

To say otherwise, and even suggest backend system remove these constraints is dangerous advice IMO.

Personally I would rather GDS come up with a set of recommended max length constraints for common fields.

E.g.

Full name -> 150
Email address -> 200
Telephone -> 15
Nat Ins. No. -> 10


I don't mind if the limits proposed are well above what would be expected, just so long as there _are_ limits.

I think there's always a need for maxlength - database integrity. While there are physical limits of what text areas/input boxes can accept (IIRC it differs between browser vendors), these limits should not be relied upon.

Similarly for file upload - users should be recommended to apply appropriate file size limits in their backend to protect against malicious users POSTing a 100GB file and bringing services down.


Many thanks

**Comment by edwardhorsford** [comment]
When I worked on this component we saw it as something intended for text areas and other *paragraph-like* responses from users. Not suitable or intended for short fields like *full name*, *email*, etc, which I'd argue is a different pattern and need.

I'd agree with @davidjamesstone that there are often cases where it's entirely appropriate to have a limit - though I'd say that's covered by the requirement that limits have a user need.

Character limits for fields where the data is of a known length is arguably a different pattern. It may be appropriate there - though you'd want to allow enough extra space to support people entering spaces or editing down - eg credit card numbers with spaces between each group of four characters.

**Comment by davidjamesstone** [comment]
Just to clear up, I wasn't suggesting this new component be used with anything other than a text area.
I just think the line:

> "Do not impose character limits unless there is a user need"

Could be misinterpreted a meaning you shouldn't use maxlength. My advice would be to _always_ use it for all input types that support it (which are inputs & textareas AFAIK). It will be one of the the first thing a penetration tester trying to break your site will look for.

> "if there is a limit set by a legacy backend system, do your best to remove the limit instead of using a character count."

I'm not sure why the word "legacy" is used here. Both new and old backend systems/database will offer some kind of support for data constraints like max field length and it's good practice to use them. Doing your best to remove them is not good advice. 
A data architect/developer will have put them in for good reasons of data quality and integrity etc. They won't be looking at it from a end user perspective). The front end should honour these constraints rather than seek to remove them.

**Comment by joelanman** [comment]
I think it's valid to say limits are needed on the back end, but that they are set to be large enough that users never normally experience them (unless there's a need to, as per the guidance).

I'm not sure it's that useful to set these large limits on the frontend, since that's no defence against a server receiving a large request - that can be done anyway, regardless of what we put in our front end. A maxlength in the frontend is a poor user experience as the user receives no error message (without javascript, and we have no-javascript progressive enhancement as a requirement). Handling it in the backend and returning an error is a better user experience.

**Comment by davidjamesstone** [comment]
> Handling it in the backend and returning an error is a better user experience.

Yeah @joelanman. Avoiding `maxlength` attributes and just relying on sensible limits in the backend error handling (using the error message + summary components) now sounds like the best overall solution to me.

**Comment by timpaul** [comment]
Thanks @davidjamesstone and @edwardhorsford for the feedback - this has been a really useful discussion. We'll review the current guidance and make sure that it's as clear as it needs to be.

**Comment by stevenaproctor** [comment]
@alex-ju I was testing this using NVDA and Firefox and noticed if you switch off JavaScript, the `aria-polite="live"` was no longer working. This means a screen reader user will only hear the 'You can enter up to 200 characters' when they leave the textarea using the down arrow. I am not sure what the experience is for other screen readers.

If we changed the `<textarea>` we could let people know the limit when they enter the textarea.

`<textarea class="govuk-textarea js-character-count " id="with-hint" name="with-hint" rows="5" aria-describedby="with-hint-hint with-hint-info"></textarea>`

This means people would hear the label, hint text and limit when entering the textarea. This should also work if JavaScript is off.

What do you think?

**Comment by alex-ju** [comment]
@stevenaproctor sounds good, thanks. will look into it and raise a PR for that.

**Comment by edwardhorsford** [comment]
I'm hoping to add this to my service shortly.

There's two things I'd like to suggest / query.

## Include a max-length for non js
Should the default html include a hardcoded limit which the js then removes? This is what we had whilst developing the service. 

A hardcoded limit is not ideal, but I think it's probably better to limit to x characters being typed at all than to let anything get typed and then throw an error. Even if the error is clear, there's no effective way for the user to cut down the text to under 200 characters.

## Allow the threshold to specify a unit in absolute characters
We originally did the threshold in percentages, but I now feel an absolute number might be better. I basically want the limit to appear _near_ the end of the field. In many cases I think this is probably easiest to specify as x characters from the end - otherwise we're going to have to vary the limit as a proportion of the total amount of characters allowed.

**Comment by stevenaproctor** [comment]
@edwardhorsford People may not notice a hardcoded limit when they are typing or pasting in content. Letting them go over the limit and giving an error would mean they have the chance to edit what they entered or paste.

**Comment by edwardhorsford** [comment]
@stevenaproctor this is only in the non-js case. In the js case, we let them go over and edit down.

It's weighing up the risk of someone pasting in and not realising it's truncated (or a AT user not knowing), or frustration of letting users type whatever, telling them it's wrong, but making them guess how much they need to remove.

Without any limit, they have to *guess* when they're at the limit - essentially delete some text, submit, delete some text, submit, delete some text, submit - a frustrating process. They have no way of knowing when they're at the limit.

**Comment by stevenaproctor** [comment]
@edwardhorsford I have seen users not realise a textarea has stopped taking what they enter or that what they have pasted is truncated. This was in a textarea without any character counter functionality and they did have to guess how much they needed to remove. I agree it is about weighing up the risk of what to do in a non-js situation.

**Comment by joelanman** [comment]
Suggestion for non-js:

 - Clear messaging up front about the character limit (this gives the user the option to use another tool to count characters, like Word)
 - No hard limit in the field - agree with @stevenaproctor that people can well not realise truncation has happened
 - Server-side validation which tells the user the content is too long, maybe message where the content passes the limit ('This goes over 200 characters at "example text here"')

**Comment by stevenaproctor** [comment]
@joelanman 👍 

Even in the js version, we do not tell people where to start in the textarea. We could definitely add a new error message to the documentation, but it could complement the 'You have 40 characters too many' error you get in the js version. Something like 'Summary has 40 characters too many'.

What do you think?

**Comment by HughePaul** [comment]
Is there a way to customise / localise the text? i.e. How do we get it in welsh?

**Comment by 36degrees** [comment]
Hi Paul,

Unfortunately customising the text used by the character count isn't currently possible. In the meantime, we suggest copying the JavaScript into your own project and making the changes to the text there.

I appreciate this isn't ideal – internationalisation / localisation is on [our roadmap](https://github.com/orgs/alphagov/projects/5) as something we want to support soon.

(This was previously raised in https://github.com/alphagov/govuk-frontend/pull/1038)

Thanks,

Ollie

**Comment by timpaul** [comment]
Related issue: https://github.com/alphagov/govuk-frontend/issues/1364

**Comment by AdamChrimes** [comment]
Hey all, we're going to be looking at implementing the Character count into the NHS Digital service manual. We're currently gathering feedback from teams using the GOV.UK character count component in NHS services. So will feedback anything we find :) 

https://github.com/nhsuk/nhsuk-service-manual-backlog/issues/172

**Comment by jamiehumphries** [comment]
> Hi Paul,
> 
> Unfortunately customising the text used by the character count isn't currently possible. In the meantime, we suggest copying the JavaScript into your own project and making the changes to the text there.
> 
> I appreciate this isn't ideal – internationalisation / localisation is on [our roadmap](https://github.com/orgs/alphagov/projects/5) as something we want to support soon.
> 
> (This was previously raised in [alphagov/govuk-frontend#1038](https://github.com/alphagov/govuk-frontend/pull/1038))
> 
> Thanks,
> 
> Ollie

@36degrees, I understand that special casing the Welsh language here is not something that you want to do because of roadmap plans for internationalisation / localisation in general.

Would you instead accept a pull request that made the character count text entirely customisable, with replacement placeholders?

I was thinking an optional property like:
```
{{ govukCharacterCount({
  …
  characterCountMessages: {
    base: "You can enter up to %count% characters",
    remaining: {
      signular: "You have 1 character remaining",
      plural: "You have %count% characters remaining"
    },
    tooMany: {
      signular: "You have 1 character too many",
      plural: "You have %count% characters too many"
    }
  }
}) }}
```

This would allow logic in Nunjucks to switch between English and Welsh, say, but without having to make Welsh an explicit special case.

If this approach sounds acceptable, I am happy to make a pull request along these lines.

**Comment by terrysimpson99** [comment]
Our use of the character count was challenged by DAC. It said:
"When navigating the confirmation page, there was a remaining character limit present underneath the ‘How can we improve this service?’ form field. When navigating the page, with the use of screen reading software the character limit was not announced to screen reader users."
[...]
“The ‘How can we improve this service’ edit field possessed a character count. This feature did not follow the GOV.UK Design System as when I entered characters, I was not given an audio prompt to alert me the character count had changed. I further found that important information such as not entering personal information was contained below the edit field and was therefore not located until after I had entered the information. The example given in the GOV.UK Design System suggests this important information should be situated between the label and the edit field. Ensuring that the character count clearly alerts me when the character count has changed, with the additional important information relating to what not to enter appears above the edit field will ensure I can access this feature easily.”

I don't know if the latest version of character count resolves this issue but we took the decision to remove the character count component.

**Comment by MalcolmVonMoJ** [comment]
Our service also has a character count (with Welsh support).  We recently passed a DAC assessment, and the character count was not flagged up as an issue.  
Our HTML is:
```
<div class="form-row ">
   <textarea class="govuk-textarea" data-character-count="4000" id="extra_notes" name="extra_notes" rows="7"></textarea>
   <div class="govuk-character-count__message govuk-hint" aria-live="polite">
      4000 nodau ar ôl
   </div>
</div>

**Comment by edwardhorsford** [comment]
@terrysimpson99 that doesn't sound like your version of the character count component was working or was using the correct component - the component was tested extensively in various AT to ensure it read out correctly. In fact in research it worked _better_ for users of screen readers than for sighted users.

**Comment by terrysimpson99** [comment]
It doesn't surprise me. The codebase was several versions behind the current one.

**Comment by terrysimpson99** [comment]
@hannalaakso In response to comments from Malcolm and Ed, please remove the 'awaiting triage' status.

**Comment by terrysimpson99** [comment]
For information:

We counted the number of characters submitted by users in one specific instance: From a survey of 10,000 instances of post-submit feedback in https://www.gov.uk/send-rent-lease-details: 99% of users entered 632 characters or fewer on the text field for 'How can we improve this service'. 99.8% entered 1000 characters or fewer.

Make of that what you will.

**Comment by SamParker1122** [comment]
> valuable found @fofr. thanks a lot for sharing it with us.
[click here](https://contadorpalavras.com/melhor-contador-de-letra/)

**Comment by ohazda** [comment]
Hey folks, awesome work on the component. I am trying to use your approach as a starting point. But unfortunately, I came across this bug or maybe feature (trying to understand why it's happening):

https://user-images.githubusercontent.com/70153084/193840386-49ad3346-80b9-47ba-826e-cabbcd4e59bf.mp4

In this particular example, the hint is getting renounced again at some point. I tried to google the issue but nothing really relates to this feature/bug. Does anybody come across it as well? Is there an explanation for this behaviour?

Tested in Chrome 105 + VO and Safari 16 + VO

I appreciate any help 🙌🏻

**Comment by stephenjmcneill1971** [comment]
I was wondering if you can help. when using the component below. 

[Character count – GOV.UK Design System (design-system.service.gov.uk)](https://design-system.service.gov.uk/components/character-count/)

A new line is only counted as one character. However when it is sent to the backend it is counted as two character CRLF instead of LF. This means the character count in the front end mis-informs the user because it says the user is within the max length when in fact they are over due to an extra character for every new line not being counted. Is there a way round this so front and back end correlate. 

I saw this has been raised before but I see no resolution. 
[Character count · Issue #172 · nhsuk/nhsuk-service-manual-community-backlog (github.com)](https://github.com/nhsuk/nhsuk-service-manual-community-backlog/issues/172)

Regards
Stephen Joseph McNeill

**Comment by davidc-gds** [comment]
> I was wondering if you can help. when using the component below.
> 
> [Character count – GOV.UK Design System (design-system.service.gov.uk)](https://design-system.service.gov.uk/components/character-count/)
> 
> A new line is only counted as one character. However when it is sent to the backend it is counted as two character CRLF instead of LF. This means the character count in the front end mis-informs the user because it says the user is within the max length when in fact they are over due to an extra character for every new line not being counted. Is there a way round this so front and back end correlate.
> 
> I saw this has been raised before but I see no resolution. [Character count · Issue #172 · nhsuk/nhsuk-service-manual-community-backlog (github.com)](https://github.com/nhsuk/nhsuk-service-manual-community-backlog/issues/172)
> 
> Regards Stephen Joseph McNeill

Hi @stephenjmcneill1971, thanks for flagging this! I don't have a quick answer for this, but it does look like this other ticket is also related to the bug you're seeing: https://github.com/alphagov/govuk-frontend/issues/1104

It might be worth it to comment there around the concern of /n being considered 2 characters in the backend, but only being valued as 1 in the frontend.

**Comment by stephenjmcneill1971** [comment]
Hi David

Thanks for updating the other ticket with my comment.

Regards
Stephen

**Comment by querkmachine** [comment]
This mismatch is, basically, the result of a difference in the specifications for HTML and HTTP. 
 
In the current specification for HTML, [new lines within a textarea are represented by a single code point](https://html.spec.whatwg.org/multipage/form-elements.html#the-textarea-element:the-textarea-element-5), LF (line feed), aka POSIX/Unix style, `\n`, or `U+000A`.
 
However, in the specification for HTTP, [new lines are represented as two code points](https://www.rfc-editor.org/rfc/rfc2616#page-16), CR (carriage return) and LF (line feed), aka Windows style, `\r\n`, or `U+000D U+000A`.
 
This is why one code point on the front-end becomes two on the back-end. It's basically 'transformed' in the HTTP transmission process.

As [our component only counts code points](https://github.com/alphagov/govuk-frontend/issues/1104), and HTML/JS will only ever see one code point, we will always have a mismatch between what we have and what the backend actually receives when it comes to new lines.

I don't think the solution is to change the character count component, however.

- We currently, strictly, only count code points. As much as this might seem weird in some circumstances (e.g. emoji), it's simple, predictable, consistent, and technically correct ([the best kind of correct](https://www.youtube.com/watch?v=0ZEuWJ4muYc)).
- Having a new line equal 1 character aligns better to a user's mental model than it equalling 2 characters.
- Ideally [we want to move to counting graphemes](https://github.com/alphagov/govuk-frontend/issues/1104#issuecomment-1333838928), rather than code points, at some point in the future. New lines aren't considered a character at all when counting graphemes, so if anything we want to move towards new lines equalling 0 characters, rather than equalling 2.
- Most modern word processors already count using graphemes and don't include new lines when reporting character counts. Ideally we want to align our behaviour as closely as possible to word processor software to avoid unpleasant surprises if a user is copying from a word processor, only to find their perfect 300 character response is suddenly 306 characters.

This does, unfortunately, leave the onus of how to validate all of this on the backend. 

As an interim solution for new lines specifically, it might be possible to detect and remove the carriage return characters from the received value before performing length validation. This would no longer suffice if we move to grapheme-based counting, however.

**Comment by frankieroberto** [comment]
Our service has a "personal statement" section where candidates have to write a longish (up to 600 words) statement about why they'd make a good teacher.

We use the [Character count component](https://design-system.service.gov.uk/components/character-count/) for this (but set to a word limit not a character limit).

However we are noticing that quite a lot of users still submit the form with the personal statement over the word limit.

Our assumptions are that one (or both) or the following might be the cause:

* they didn't spot the as-you-type word count – maybe because they pasted the content from elsewhere?
* they know it's over the word limit but want to save it and edit it down later

We’re currently exploring ways to potentially mitigate this.

**Comment by stevenjmesser** [comment]
Sounds good, @frankieroberto, let us know what you find!

**Comment by andythenorth** [comment]
Hi (also hi @stevenjmesser ) we're not Design System users here at Delib, but we use it as reference for Citizen Space UI, we have recently run into the same issues, and have had a crash course in graphemes.  Watching with interest.

**Comment by L-e-a-n-n-e-H** [comment]
We’ve come across an issue whereby when someone types into the text area box if there is a delay of about 1 second before they type another character in, the screen reader will read out the remaining character count, this continues to happen for the entire sentence they are writing. 
This can be a bit excessive for people who do not type or can type quickly and use a screen reader.  Has anyone set the character count to more than 1 second, is this achievable? Or does anyone know of an optimal length of time before the character count starts to read out on a screen reader? Thank you :)

**Comment by davidc-gds** [comment]
Hi @L-e-a-n-n-e-H, thanks for providing this feedback! I have a couple clarifying questions on the specific scenario, if you're able to provide more info. It's alright if you don't have answers for all the questions though. 👍

- Which screen reader was in use, with which browser, and on what operating system / device?
- Do you know approximately how fast the person was typing when the encountered the repeated character count announcements? The 1 second delay timer was based on the assumption that in a normal typing scenario, the typing speed would be greater than 60 characters per second (the general rule-of-thumb average typing speed is around 190 characters per second, more than 3 times quicker).
- Do you know what version of GOV.UK Frontend is being used? The screen reader announcements were much more intrusive and frequent before we made the updates. The updated character count component was released in version 4.1.0 in May 2022.
- Is the person pausing after typing each character, or is the issue that they're typing long strings of characters and the screen reader is announcing the character count message while they're still typing at a normal pace?

**Comment by L-e-a-n-n-e-H** [comment]
Hi @davidc-gds , many thanks for the quick reply, much appreciated!

- We had an accessibility audit that picked it up on the web browser of the NHS App (desktop) they were using a Mac, with NVDA on Firefox.
I’ve tried with the Chrome extension and it does the same. 

- And relates to the text box prior to booking a GP appointment, in a texarea box that is mandatory.

- The person would have long pauses, and is most likely with someone who needs to take their time; has dexterity issues, unable to touch type, needs longer processing time to move between letters/characters and articulate what they want to get across (we're awaiting data/analytics to probe further). 

 In the meantime we were looking at what would be a good 'pause' length from the current 1000ms (1 second) and if anyone had any suggestions for this?

Thank you!

![characters remaining ](https://user-images.githubusercontent.com/112878173/230386934-01169f12-3167-4d8c-ba60-6d49d1832f4e.png)

**Comment by CharlotteDowns** [comment]
We ran an external accessibility audit for some of the components and patterns in GOV.UK Frontend in May 2023. In that audit, we included an example of the Character count component. We’re adding results from that audit here so that we can:
- document, discuss and address the findings
- inform future contributors of the findings

## One usability accessibility issue raised

The hidden character count only pops up when you approach threshold is strange usability-wise.
> The character count for the text area ‘What were you trying to do?’ does not appear until the user reaches fifty characters. It can be beneficial to display this from the start and the max characters to allow users to understand this restriction before entering data.
This means that when the user reaches 50 characters, and are made aware, they then do not have to amend what they have already entered in order to not exceed the 100 limit.

More detail in this issue:
- [#3689](https://github.com/alphagov/govuk-frontend/issues/3689)

**Comment by joelanman** [comment]
Just to note the threshold is an optional feature. We designed it with the aim that most fields would be large enough to accept most input, and not make the user have to think about the character/word limit. In the minority case where users might hit the limit, the threshold option allows those users to know they are near the limit.

**Comment by RobSinclair** [comment]
Feedback from a DAC Accessibility Audit: 
![image](https://github.com/user-attachments/assets/e5c52ba9-d5bf-442a-ae07-2fd9752fa138)

Following this feedback we will be removing the aria-hidden attribute and including a status="polite" attribute so that the updating status does not interrupt the feedback when typing, and provides the remaining character count at the end of audio feedback for screen reader users. 

Curious as to if anyone else has done similar and how that has tested?

**Comment by DeanRITS** [comment]
Hi,

I can see on the nunjucks macro options that it allows wording to be passed to the component using the following options:

charactersUnderLimitText
charactersOverLimitText
charactersAtLimitText

Unfortunately I'm not using nunjucks so I am attempting to do this via html with data attributes. I've been able to override some of the wording using the following attributes:

data-i18n.characters-under-limit.other="custom wording" 
data-i18n.characters-under-limit.one="custom wording" 
data-i18n.characters-over-limit.other="custom wording"
data-i18n.characters-over-limit.one="custom wording" 

But I cannot get the message for charactersAtLimitText to override. Any suggestions?

Thanks,
Dean

**Comment by ijfessey** [comment]
Hi team, We are potentially going to use this component in the self-assesment application and we have it working but the question is can the 'you have n characters remaining' be overridden as we need to support Welsh language ? 

<img width="500" alt="Image" src="https://github.com/user-attachments/assets/ec73f2cd-0990-49e0-a0e9-f0e9b7a04ded" />

**Comment by RichardBradley** [comment]
@ijfessey -- here's a copy of our Welsh localisation wrapper around this component https://gist.github.com/RichardBradley/70dfa1cae92ff3d2c7aa9377eb6d4d8e

I think GDS ought to add this as an enhancement, many gov projects require Welsh. If I get some time, I'll raise a PR against https://github.com/alphagov/govuk-frontend based on the above
